resource/logging_macros.h.em

```diff
@@ -0,0 +1,414 @@
+
+/** @@name Macros for the `expression` condition which ignores the log calls
+ * when the expression evaluates to false.
+ */
+///@@{
+/**
+ * \def RCUTILS_LOG_CONDITION_EXPRESSION_BEFORE
+ * A macro checking the `expression` condition.
+ */
+#define RCUTILS_LOG_CONDITION_EXPRESSION_BEFORE(expression) \
+  if (expression) {
+/**
+ * \def RCUTILS_LOG_CONDITION_EXPRESSION_AFTER
+ * A macro finalizing the `expression` condition.
+ */
+#define RCUTILS_LOG_CONDITION_EXPRESSION_AFTER }
+///@@}
+
+/** @@name Macros for the `function` condition which ignores the log calls
+ * when the function returns false.
+ */
+///@@{
+/// The filter function signature.
+/**
+ * \return true to log the message, false to ignore the message
+ */
+typedef bool (* RclLogFilter)();
+/**
+ * \def RCUTILS_LOG_CONDITION_FUNCTION_BEFORE
+ * A macro checking the `function` condition.
+ */
+#define RCUTILS_LOG_CONDITION_FUNCTION_BEFORE(function) \
+  if ((*function)()) {
+/**
+ * \def RCUTILS_LOG_CONDITION_FUNCTION_AFTER
+ * A macro finalizing the `function` condition.
+ */
+#define RCUTILS_LOG_CONDITION_FUNCTION_AFTER }
+///@@}
+
+/** @@name Macros for the `skipfirst` condition which ignores the first log
+ * call but processes all subsequent calls.
+ */
+///@@{
+/**
+ * \def RCUTILS_LOG_CONDITION_SKIPFIRST_BEFORE
+ * A macro initializing and checking the `skipfirst` condition.
+ */
+#define RCUTILS_LOG_CONDITION_SKIPFIRST_BEFORE \
+  { \
+    static bool __rcutils_logging_first = true; \
+    if (RCUTILS_UNLIKELY(true == __rcutils_logging_first)) { \
+      __rcutils_logging_first = false; \
+    } else {
+/**
+ * \def RCUTILS_LOG_CONDITION_SKIPFIRST_AFTER
+ * A macro finalizing the `skipfirst` condition.
+ */
+#define RCUTILS_LOG_CONDITION_SKIPFIRST_AFTER } \
+  }
+///@@}
+
+/** @@name Macros for the `throttle` condition which ignores log calls if the
+ * last logged message is not longer ago than the specified duration.
+ */
+///@@{
+/**
+ * \def RCUTILS_LOG_CONDITION_THROTTLE_BEFORE
+ * A macro initializing and checking the `throttle` condition.
+ */
+#define RCUTILS_LOG_CONDITION_THROTTLE_BEFORE(time_source_type, duration) { \
+    static rcutils_duration_value_t __rcutils_logging_duration = RCUTILS_MS_TO_NS((rcutils_duration_value_t)duration); \
+    static rcutils_time_point_value_t __rcutils_logging_last_logged = 0; \
+    rcutils_time_point_value_t __rcutils_logging_now = 0; \
+    bool __rcutils_logging_condition = true; \
+    if (rcutils_steady_time_now(&__rcutils_logging_now) != RCUTILS_RET_OK) { \
+      rcl_log( \
+        &__rcutils_logging_location, RCUTILS_LOG_SEVERITY_ERROR, "", \
+        "%s() at %s:%d getting current steady time failed\n", \
+        __func__, __FILE__, __LINE__); \
+    } else { \
+      __rcutils_logging_condition = __rcutils_logging_now >= __rcutils_logging_last_logged + __rcutils_logging_duration; \
+    } \
+ \
+    if (RCUTILS_LIKELY(__rcutils_logging_condition)) { \
+      __rcutils_logging_last_logged = __rcutils_logging_now;
+
```

Comment:
Oh, I see now.  Because this is the "before" condition, when we generate the print line we are still inside of the `if` condition.  That detail is a bit subtle; it might be worthwhile to put a comment explaining that.